Shrink dist-repo-status Table to Terminal Width

jmgate opened the pull request:
If necessary, shrink the `dist-repo-status` table to the width of the terminal by shrinking the `Repo Dir`, `Branch`, and `Tracking Branch` columns.  Shrink labels/fields by inserting `...` in the middle of them.

jmgate commented:
@bartlettroscoe, this hopefully won't run into any version compatibility issues&mdash;fingers crossed.

All existing unit tests pass, but I haven't written any to test the shrinking of the table in particular&mdash;only tested it out by running `dist-repo-status` in my own terminal.

**Note:**  This may run into merge conflicts with #255.

bartlettroscoe commented:
@jmgate, I tried this out for a shorter terminal width and it showed:

```
---------------------------------------------------------------------
| ID | Repo ...o Dir | Bra...nch | Tracking Branch                                  | C | M | ? |
|----|||--------------------------------------------------|---|---|---|
|  0 | Trilinos (Ba...linos (Base) | 2462-openmpi-1.6.5-to-1.1...2-openmpi-1.6.5-to-1.10.1 | rab-github/2462-openmpi-1.6.5-to-1.10.1          |   |   |   |
|  1 | TriB...BITS | shrink-table-to-fit-termi...ink-table-to-fit-terminal | jmgate/shrink-table-to-fit-terminal              |   |   |   |
|  2 | TriBITS/TriBITS...BITS/TriBITSDoc | mas...ter | bb/master                                        |   |   |   |
|  3 | kok...kos | deve...elop | github/develop                                   |   |   |   |
|  4 | kokkos-kern...kos-kernels | deve...elop | github/develop                                   |   |   |   |
|  5 | sea...cas | mas...ter | gsjaardema/master                                |   |   |   |
---------------------------------------------------------------------
```

Is that the desired look of the output?  Should not the columns marks all line up?  Also, the fields in the first column are actually longer with incorrect text.  For example `Trilinos (Ba...linos (Base)` is longer than the correct value `Trilinos (Base)`.

I tried the same command with the current implementation and it showed:

![gitdist-dist-repo-stauts-narrow-terminal](https://user-images.githubusercontent.com/1959736/38087786-7474931c-3327-11e8-84a7-64d905a57f9b.png)

That looks more readable to me.

Also, are there any tests to demonstrate an protect this new feature?

jmgate commented:
Nope, that looks like garbage.  Here's what I'm seeing:
```
-------------------------------------------------------------------------------------
| ID | Repo Dir  | Branch              | Tracking Branch                | C | M | ? |
|----|-----------|---------------------|--------------------------------|---|---|---|
|  0 | Tri...se) | 2462-ope...o-1.10.1 | rab-github/246...6.5-to-1.10.1 |   |   |   |
|  1 | TriBITS   | shrink-t...terminal | jmgate/shrink-...-fit-terminal |   |   |   |
|  2 | kokkos    | develop             | github/develop                 |   |   |   |
|  3 | kok...els | develop             | github/develop                 |   |   |   |
|  4 | seacas    | master              | gsjaardema/master              |   |   |   |
-------------------------------------------------------------------------------------
```

Debugging&hellip;

jmgate commented:
Blast!  Silly integer division.  Care to try again?

There are no unit tests for this shrinking ability yet&mdash;trying to figure out how to control terminal size via Python&hellip;

bartlettroscoe commented:
> Blast! Silly integer division. Care to try again?

Pulled the updated branch and tried again with a wide terminal:

```
----------------------------------------------------------------------------------------------------------------
| ID | Repo Dir           | Branch                       | Tracking Branch                         | C | M | ? |
|----|--------------------|------------------------------|-----------------------------------------|---|---|---|
|  0 | Trilinos (Base)    | 2462-openmpi-1.6.5-to-1.10.1 | rab-github/2462-openmpi-1.6.5-to-1.10.1 |   |   |   |
|  1 | TriBITS            | shrink-table-to-fit-terminal | jmgate/shrink-table-to-fit-terminal     |   |   |   |
|  2 | TriBITS/TriBITSDoc | master                       | bb/master                               |   |   |   |
|  3 | kokkos             | develop                      | github/develop                          |   |   |   |
|  4 | kokkos-kernels     | develop                      | github/develop                          |   |   |   |
|  5 | seacas             | master                       | gsjaardema/master                       |   |   |   |
----------------------------------------------------------------------------------------------------------------
```

and a more narrow terminal where the full table does not fit.

```
---------------------------------------------------------------------------
| ID | Repo Dir   | Branch          | Tracking Branch         | C | M | ? |
|----|------------|-----------------|-------------------------|---|---|---|
|  0 | Tril...se) | 2462-o...1.10.1 | rab-github...-to-1.10.1 |   |   |   |
|  1 | TriBITS    | shrink...rminal | jmgate/shr...t-terminal |   |   |   |
|  2 | TriB...Doc | master          | bb/master               |   |   |   |
|  3 | kokkos     | develop         | github/develop          |   |   |   |
|  4 | kokk...els | develop         | github/develop          |   |   |   |
|  5 | seacas     | master          | gsjaardema/master       |   |   |   |
---------------------------------------------------------------------------
```

That looks nice!

> There are no unit tests for this shrinking ability yet—trying to figure out how to control terminal size via Python…

What you can do is to use an env var to provide you the value returned from `os.popen("stty size", "r").read()`.  For example, replace the line of code:

```
rows, columns = os.popen("stty size", "r").read().split()
```

with:

```
mockSttySize = os.environ.get("GITDIST_UNIT_TEST_STTY_SIZE")

if mockSttySize:
   sttySize = mockSttySize
else:
  sttySize = os.popen("stty size", "r").read()

rows, columns = sttySize.split()
```

then in your unit tests, set the env var `GITDIST_UNIT_TEST_STTY_SIZE` for different terminal sizes and then assert that the table looks correct.

The unit tests should run at least these 5 use cases:

1. The terminal size is larger than needed for the full table
2. The terminal size will exactly fit in the full table
3. The terminal size is smaller than will fit in the table (but not my much)
4. The terminal size is very small and therefore has to be compress a lot
5. The terminal size is too small even for the column names

I just tried the last case manually and it returned:

```
-----------------------------------
| ID | Repo Di...epo Dir | ...ranch | T...h | C | M | ? |
|----|---|----|-------|---|---|---|
|  0 | Trilinos (Base...rilinos (Base) | ...462-openmpi-1.6.5-to-1.10.1 | r...1 |   |   |   |
|  1 | TriBIT...riBITS | ...hrink-table-to-fit-terminal | j...l |   |   |   |
|  2 | TriBITS/TriBITSDo...riBITS/TriBITSDoc | ...aster | b...r |   |   |   |
|  3 | kokko...okkos | ...evelop | g...p |   |   |   |
|  4 | kokkos-kernel...okkos-kernels | ...evelop | g...p |   |   |   |
|  5 | seaca...eacas | ...aster | g...r |   |   |   |
-----------------------------------
```

That just looks like gobbledygook.  It seems like the min width for each column should be the column titles.  If the table is any smaller than than, then don't bother to try to shrink it more.  Does that sound like a reasonable rule?

Thanks for working on this!

jmgate commented:
Sounds like a good idea.  I may have some time tomorrow to work in those changes.

bartlettroscoe commented:
@jmgate, have you given up on this PR?

jmgate commented:
Haven't given up, just haven't had free time.  Might be a good way to start a Friday morning though&hellip;

jmgate commented:
Now the table won't shrink if doing so would have caused the abbreviation of any column heading.  If you're in such a situation, your terminal window is just too narrow.

Also added the following to allow for testing:
```
mockSttySize = os.environ.get("GITDIST_UNIT_TEST_STTY_SIZE")
if mockSttySize:
   sttySize = mockSttySize
else:
  sttySize = os.popen("stty size", "r").read()
rows, columns = sttySize.split()
```

Thanks @bartlettroscoe.  Now on to unit testing&hellip;

jmgate commented:
@bartlettroscoe, you want to force pull and take a look at these unit tests?  I think this is probably ready to go, unless something's wonky with Python 2, but I ran gitdist_UnitTests.py via RHEL7's Python 2.7.5 and these tests looked fine.

jmgate commented:
Note that this will conflict with #255.  Whichever one gets merged first, I can handle the merge conflicts and update the remaining one.

bartlettroscoe commented:
@jmgate, I will take a look once I get access back to my SNL Linux machine (NX server is down).

bartlettroscoe commented:
@jmgate, I tested this with:

```
$ python --version
Python 2.6.6
```

and it seems to be shrinking the table nicely as the terminal gets smaller.  However, when the terminal gets small enough, it prints out the full table and just wraps the line like:

```
-------------------------------------------------------------------------------------------------------------
| ID | Repo Dir            | Branch                       | Tracking Branch                     | C | M | ? |
|----|---------------------|------------------------------|-------------------------------------|---|---|---|
|  0 | Trilinos (Base)     | develop                      | github/develop                      |   |   |   |
|  1 | TriBITS             | shrink-table-to-fit-terminal | jmgate/shrink-table-to-fit-terminal |   |   |   |
|  2 | TriBITS/TriBITSDoc  | master                       | bb/master                           |   |   |   |
|  3 | kokkos              | develop                      | github/develop                      |   |   |   |
|  4 | kokkos/nvcc_wrapper | develop                      | github/develop                      |   |   |   |
|  5 | kokkos-kernels      | develop                      | github/develop                      |   |   |   |
|  6 | seacas              | master                       | gsjaardema/master                   |   |   |   |
-------------------------------------------------------------------------------------------------------------
```

But just before it gets to that smaller width, it prints out some pretty reasonable tables like:

```
-----------------------------------------------------------------
| ID | Repo Dir | Branch       | Tracking Branch    | C | M | ? |
|----|----------|--------------|--------------------|---|---|---|
|  0 | Tri...e) | develop      | github/develop     |   |   |   |
|  1 | TriBITS  | shrin...inal | jmgate/s...erminal |   |   |   |
|  2 | Tri...oc | master       | bb/master          |   |   |   |
|  3 | kokkos   | develop      | github/develop     |   |   |   |
|  4 | kok...er | develop      | github/develop     |   |   |   |
|  5 | kok...ls | develop      | github/develop     |   |   |   |
|  6 | seacas   | master       | gsjaardema/master  |   |   |   |
-----------------------------------------------------------------
```

and

```
----------------------------------------------------------------------------
| ID | Repo Dir    | Branch           | Tracking Branch        | C | M | ? |
|----|-------------|------------------|------------------------|---|---|---|
|  0 | Tril...ase) | develop          | github/develop         |   |   |   |
|  1 | TriBITS     | shrink-...rminal | jmgate/shr...-terminal |   |   |   |
|  2 | TriB...SDoc | master           | bb/master              |   |   |   |
|  3 | kokkos      | develop          | github/develop         |   |   |   |
|  4 | kokk...pper | develop          | github/develop         |   |   |   |
|  5 | kokk...nels | develop          | github/develop         |   |   |   |
|  6 | seacas      | master           | gsjaardema/master      |   |   |   |
----------------------------------------------------------------------------
```

and

```
----------------------------------------------------------------------------------------
| ID | Repo Dir       | Branch               | Tracking Branch             | C | M | ? |
|----|----------------|----------------------|-----------------------------|---|---|---|
|  0 | Trilin...Base) | develop              | github/develop              |   |   |   |
|  1 | TriBITS        | shrink-ta...terminal | jmgate/shrin...fit-terminal |   |   |   |
|  2 | TriBIT...TSDoc | master               | bb/master                   |   |   |   |
|  3 | kokkos         | develop              | github/develop              |   |   |   |
|  4 | kokkos...apper | develop              | github/develop              |   |   |   |
|  5 | kokkos-kernels | develop              | github/develop              |   |   |   |
|  6 | seacas         | master               | gsjaardema/master           |   |   |   |
----------------------------------------------------------------------------------------
```

and

```
----------------------------------------------------------------------------------------------
| ID | Repo Dir        | Branch                 | Tracking Branch                | C | M | ? |
|----|-----------------|------------------------|--------------------------------|---|---|---|
|  0 | Trilinos (Base) | develop                | github/develop                 |   |   |   |
|  1 | TriBITS         | shrink-tab...-terminal | jmgate/shrink-...-fit-terminal |   |   |   |
|  2 | TriBIT...ITSDoc | master                 | bb/master                      |   |   |   |
|  3 | kokkos          | develop                | github/develop                 |   |   |   |
|  4 | kokkos...rapper | develop                | github/develop                 |   |   |   |
|  5 | kokkos-kernels  | develop                | github/develop                 |   |   |   |
|  6 | seacas          | master                 | gsjaardema/master              |   |   |   |
----------------------------------------------------------------------------------------------
```

and

```
-----------------------------------------------------------------------------------------------------------
| ID | Repo Dir           | Branch                      | Tracking Branch                     | C | M | ? |
|----|--------------------|-----------------------------|-------------------------------------|---|---|---|
|  0 | Trilinos (Base)    | develop                     | github/develop                      |   |   |   |
|  1 | TriBITS            | shrink-table...fit-terminal | jmgate/shrink-table-to-fit-terminal |   |   |   |
|  2 | TriBITS/TriBITSDoc | master                      | bb/master                           |   |   |   |
|  3 | kokkos             | develop                     | github/develop                      |   |   |   |
|  4 | kokkos/n...wrapper | develop                     | github/develop                      |   |   |   |
|  5 | kokkos-kernels     | develop                     | github/develop                      |   |   |   |
|  6 | seacas             | master                      | gsjaardema/master                   |   |   |   |
-----------------------------------------------------------------------------------------------------------
```

And no matter what the terminal size is, when you pipe to a file like:

```
$ gitdist dist-repo-status &> gitdist.out
```

it always (correctly) prints the full table.

This looks very nice.

bartlettroscoe commented on test/python_utils/gitdist_UnitTests.py:
```diff
@@ -196,6 +200,112 @@ def test_row_mismatch(self):
+    self.assertEqual(asciiTable, asciiTable_expected)
+
+
+  def test_stty_size_too_small_for_heading(self):
```
Okay, and here is the test for the behavior when the terminal width is too small to fit all of the headers and it prints the full table.  Nice!

bartlettroscoe commented on tribits/python_utils/gitdist.py:
```diff
@@ -722,6 +722,17 @@ def getHelpTopicsStr():
 
+# Shrink a string to a given width by inserting an ellipsis (...) in the
+# middle.
+def shrinkString(string, width):
```
We could unit test this function as well.  But the upper-level table shrinking tests test this.

bartlettroscoe commented:
@jmgate, if you want to write unit tests for `shrinkString()`, then we can wait to merge.  Otherwise, it would be easy to add unit tests for this function in case we want to edit what it does later.

jmgate commented:
Let's go ahead and merge now.  I'm afraid I should be working on other things, but whenever I come back to hack around with gitdist, I'll keep this in mind.

bartlettroscoe commented:
@jmgate, not sure how this happened, but the merge of this branch resulted in test failures for the Travis CI build as shown on CDash at:

* https://my.cdash.org/viewTest.php?onlyfailed&buildid=1492294

I thought that Travis CI said it has passed?  I wonder if Travis CI tests the merge of your topic branch into  'master' first or just tests the topic branch in isolation?  That might explain how this might happen if the branch was created off of an other version of the TriBITS repo.

jmgate commented:
That's bizarre.  I thought Travis tested the merge, but either way this branch was a fast-forward.  It looks like Travis is setting the terminal size to 80 characters wide somehow.  You want to revert?

jmgate commented:
Alternatively, take a look at the head of jmgate/TriBITS:shrink-table-to-fit-terminal (65ef1f5)&mdash;that should fix things for Travis.  I didn't put these bits in earlier because it looked like they weren't needed.

bartlettroscoe commented:
Can we create a PR and let Travis CI test this?
